### PR TITLE
General index for microphysics flux

### PR DESCRIPTION
**Pull request by ChunYen-Chen**
## Major
* Add a microphysics flux template.
* Add general function for viscosity and conduction. (`Hydro_AddViscousFlux()` and `Hydro_AddConductiveFlux()`)
* Fix some compilation bugs.

## Minor
* Remove trailing blank.
* Update comments.

### CODE REVIEW

**Comment by jzuhone** [comment]
@ChunYen-Chen coming back to this rather late, but do you remember why we need to call `Hydro_AddConductiveFlux` twice, in `CUFLU_FluidSolver_MHM` and `Hydro_DataReconstruction`?

**Comment by ChunYen-Chen** [comment]
@jzuhone

## `MHM` scheme
The half-step update is done in `Hydro_Predict()` ([here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp#L2228)) after the data-reconstruction. Thus, the additional flux of the half-step needs to be calculated before `Hydro_Predict()` ([here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp#L420)). The full-step update is done in `Hydro_FullStepUpdate()` ([here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp#L693)), so the additional full-step flux is calculated before the full-step update.

## `MHM_RP` scheme
The half-step update is done in `Hydro_RiemannPredict()` ([here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp#L565)), so the additional half-step flux calculation is needed before the half-step update ([here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp#L540)). The full-step update is shared with the `MHM` scheme.

## Bug
I found that there is a [bug](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Microphysics/Conduction/CPU_AddConductiveFlux.cpp#L778-L783) in the `Hydro_AddConductiveFlux()` and `Hydro_AddViscousFlux()`. This bug can only be triggered by the `MHM` scheme which I did not test. 😢 The `initialize` variable should be updated to `false` after updating all the fluxes instead of one flux.
To fix this bug, please remove the `initialize = false;` [here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Microphysics/Conduction/CPU_AddConductiveFlux.cpp#L782) and add `if ( initialize ) initialize = false;` [here](https://github.com/jzuhone/gamer/blob/558f5f3aae58f4ff7e05c3149e4092d584c722e2/src/Microphysics/Conduction/CPU_AddConductiveFlux.cpp#L790). The final code should be like:
```c++
GPU_DEVICE
void Hydro_AddConductiveFlux( ... )
{
   ...
   for (int d=0; d<3; d++)
   {
      ...

      CGPU_LOOP( idx, flux_size_i*flux_size_j*flux_size_k )
      {
         ...

//       5. initialize flux if need
         if ( initialize )
         {
            for (int v=0; v<NCOMP_TOTAL_PLUS_MAG; v++)   g_Flux[d][v][idx_flux] = (real)0.0;
         }

         ...
      } // CGPU_LOOP( idx, flux_size_i*flux_size_j*flux_size_k )
   } // for (int d=0; d<3; d++)

   if ( initialize ) initialize = false;

#  ifdef __CUDACC__
   __syncthreads();
#  endif

} // FUNCTION : Hydro_AddConductiveFlux
```
Would you like me to file a PR to fix this bug?

File: src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp [hunk 1]
```diff
@@ -654,8 +745,11 @@ void Hydro_DataReconstruction( const real g_ConVar   [][ CUBE(FLU_NXT) ],
 
 
 #     if ( FLU_SCHEME == MHM )
+      real Extra_Flux[6][NCOMP_TOTAL_PLUS_MAG] = {0.0};
+      for (int f=0; f<6; f++)
```
Comment (jzuhone) [inline src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:749]: @ChunYen-Chen this looks like some unnecessary lines of code. Can it be deleted or were we supposed to use them?

File: src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp [hunk 2]
```diff
@@ -543,10 +587,10 @@ void CPU_FluidSolver_MHM(
 
 //          1-a-5. evaluate the face-centered values by data reconstruction
 //                 --> note that g_PriVar_Half_1PG[] returned by Hydro_RiemannPredict() stores the primitive variables
-            Hydro_DataReconstruction( NULL, g_FC_Mag_Half_1PG, g_PriVar_Half_1PG, g_FC_Var_1PG, g_Slope_PPM_1PG,
+            Hydro_DataReconstruction( NULL, g_FC_Mag_Half_1PG, g_PriVar_Half_1PG, g_FC_Var_1PG, NULL, g_Slope_PPM_1PG,
```
Comment (jzuhone) [inline src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp:590]: @ChunYen-Chen should `g_Flux` be `NULL` here?

**Comment by jzuhone** [comment]
Hi @ChunYen-Chen:

> Would you like me to file a PR to fix this bug?

No worries, I've fixed it. I have a couple other questions that I have put in. Could you check? I'll have one other question later.

File: src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp (hunk 1 above)
Comment (ChunYen-Chen) [reply src/Model_Hydro/CPU_Hydro/CPU_Shared_DataReconstruction.cpp:749]: Yes, you can remove `Extra_Flux`.

File: src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp (hunk 2 above)
Comment (ChunYen-Chen) [reply src/Model_Hydro/CPU_Hydro/CPU_FluidSolver_MHM.cpp:590]: Yes, the half-step update is done by `Hydro_RiemannPredict()` for the `MHM_RP` scheme.

**Comment by ChunYen-Chen** [comment]
@jzuhone Just in case you did not receive the notification. I have left the comments in the conversation.

**Pull request merged**